polkadot/node/core/pvf/common/src/worker/security/seccomp.rs

```diff
@@ -0,0 +1,197 @@
+//!
+//! Considering how many things would have to go wrong here, we believe it's safe to block
+//! `io_uring`.
+//!
+//! # Action on	caught syscall
+//!
+//!	TODO
+
+use crate::{
+	worker::{stringify_panic_payload, WorkerKind},
+	LOG_TARGET,
+};
+use seccompiler::*;
+use std::{collections::BTreeMap, path::Path};
+
+/// The action to take on caught syscalls.
+#[cfg(not(test))]
+const CAUGHT_ACTION: SeccompAction = SeccompAction::KillProcess;
+/// Don't kill the process when testing.
+#[cfg(test)]
+const CAUGHT_ACTION: SeccompAction = SeccompAction::Errno(libc::EACCES as u32);
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+	#[error(transparent)]
+	Seccomp(#[from] seccompiler::Error),
+	#[error(transparent)]
+	Backend(#[from] seccompiler::BackendError),
+	#[error("A panic occurred in try_restrict: {0}")]
+	Panic(String),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Try to enable seccomp for the given kind of worker.
+pub fn enable_for_worker(
+	worker_kind: WorkerKind,
+	worker_pid: u32,
+	worker_dir_path: &Path,
+) -> Result<()> {
+	gum::trace!(
+		target: LOG_TARGET,
+		%worker_kind,
+		%worker_pid,
+		?worker_dir_path,
+		"enabling seccomp",
+	);
+
+	try_restrict()
+}
+
+/// Runs a check for seccomp and returns a single bool indicating whether seccomp with our rules is
+/// fully enabled on the current Linux environment.
+pub fn check_is_fully_enabled() -> bool {
+	let status_from_thread: Result<()> = match std::thread::spawn(|| try_restrict()).join() {
+		Ok(Ok(())) => Ok(()),
+		Ok(Err(err)) => Err(err.into()),
+		Err(err) => Err(Error::Panic(stringify_panic_payload(err))),
+	};
+
+	matches!(status_from_thread, Ok(()))
+}
+
+/// Applies a `seccomp` filter to disable networking for the PVF threads.
+pub fn try_restrict() -> Result<()> {
+	// Build a `seccomp` filter which by default allows all syscalls except those blocked in the
+	// blacklist.
+	let mut blacklisted_rules = BTreeMap::default();
+
+	// Restrict the creation of sockets.
+	blacklisted_rules.insert(libc::SYS_socketpair, vec![]);
+	blacklisted_rules.insert(libc::SYS_socket, vec![]);
+
+	// Restrict io_uring.
+	blacklisted_rules.insert(libc::SYS_io_uring_setup, vec![]);
+	blacklisted_rules.insert(libc::SYS_io_uring_enter, vec![]);
+	blacklisted_rules.insert(libc::SYS_io_uring_register, vec![]);
+
+	// Restrict connecting to abstract unix sockets.
+	blacklisted_rules.insert(libc::SYS_connect, vec![]);
+
+	let filter = SeccompFilter::new(
+		blacklisted_rules,
+		// Mismatch action: what to do if not in rule list.
+		SeccompAction::Allow,
+		// Match action: what to do if in rule list.
+		CAUGHT_ACTION,
+		TargetArch::x86_64,
+	)?;
+
+	let bpf_prog: BpfProgram = filter.try_into()?;
+
+	// Applies filter (runs seccomp) to the calling thread.
+	seccompiler::apply_filter(&bpf_prog)?;
+
+	Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use std::{io::ErrorKind, net::TcpListener, thread};
+
+	#[test]
+	fn sandboxed_thread_cannot_use_sockets() {
```

Comment:
shouldn't we only run this test on x86_64?

Reply:
The whole module is gated on `x86_64`. As an aside, there should be a compile error on Linux arm64. I'm waiting to see if CI catches it because I have no way of testing that myself.

Reply:
Looks like we have no CI job for arm64. 🤔

Reply:
> The whole module is gated on x86_64. As an aside, there should be a compile error on Linux arm64. I'm waiting to see if CI catches it because I have no way of testing that myself.

Didn't notice that 👍🏻 

yes, the current code should fail to compile on arm. if you have an Arm macbook you can cross compile for linux like: `cargo build --target aarch64-unknown-linux-gnu` (and `rustup target add ...` beforehand)

Reply:
Thanks, caught the error with your help and https://github.com/messense/homebrew-macos-cross-toolchains and fixed it.